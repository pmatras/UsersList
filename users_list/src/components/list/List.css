ol {
  color: #63625c;
}

li + li {
  margin-top: 10px;
}
