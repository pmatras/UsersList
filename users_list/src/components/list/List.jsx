import React from 'react';
import PropTypes from 'prop-types';

import './List.css';

const defaultRenderFunction = (item) => <span>{JSON.stringify(item)}</span>;

const List = ({
  items,
  noItemsMsg = 'No elements',
  renderFunction,
  itemId = 'id',
}) => {
  return items?.length ? (
    <div data-testid="list-container">
      <ol data-testid="list">
        {items.map((item) => (
          <li key={item[itemId]} data-testid={`li-${item[itemId]}`}>
            {renderFunction?.(item) || defaultRenderFunction(item)}
          </li>
        ))}
      </ol>
    </div>
  ) : (
    <div data-testid="no-items-container">{noItemsMsg}</div>
  );
};

List.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    })
  ).isRequired,
  noItemsMsg: PropTypes.string,
  renderFunction: PropTypes.func,
  itemId: PropTypes.string,
};

export default List;
