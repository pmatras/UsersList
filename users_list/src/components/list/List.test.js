import React from 'react';
import { render, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';

import List from './List';

afterAll(cleanup);

describe('List component tests', () => {
  it('renders successfully when empty items list passed', () => {
    const { getByTestId } = render(<List items={[]} />);
    expect(getByTestId('no-items-container')).toBeInTheDocument();
    expect(getByTestId('no-items-container')).toHaveTextContent('No elements');
  });

  it('renders successfully when non empty items list passed', () => {
    const items = [
      {
        id: 0,
        name: 'item1',
      },
      {
        id: 1,
        name: 'item2',
      },
    ];
    const { getByTestId } = render(<List items={items} />);
    expect(getByTestId('list-container')).toBeInTheDocument();
    expect(getByTestId('list-container')).toHaveTextContent(
      `${JSON.stringify(items[0])}${JSON.stringify(items[1])}`
    );
  });

  it('uses properly custom render function', () => {
    const items = [
      {
        id: 0,
        name: 'item1',
      },
      {
        id: 1,
        name: 'item2',
      },
    ];
    const { getByTestId } = render(
      <List items={items} renderFunction={({ name }) => <span>{name}</span>} />
    );
    expect(getByTestId('list-container')).toHaveTextContent(
      `${items[0].name}${items[1].name}`
    );
  });

  it('uses properly custom no items message', () => {
    const noItemsMsg = 'None element found';
    const { getByTestId } = render(<List items={[]} noItemsMsg={noItemsMsg} />);
    expect(getByTestId('no-items-container')).toHaveTextContent(noItemsMsg);
  });

  it('matches snapshot with mocked elements list', () => {
    const items = [
      {
        id: 0,
        name: 'item1',
      },
      {
        id: 1,
        name: 'item2',
      },
    ];
    const tree = renderer.create(<List items={items} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with empty elements list', () => {
    const tree = renderer.create(<List items={[]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
