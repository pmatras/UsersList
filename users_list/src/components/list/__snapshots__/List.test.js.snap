// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component tests matches snapshot with empty elements list 1`] = `
<div
  data-testid="no-items-container"
>
  No elements
</div>
`;

exports[`List component tests matches snapshot with mocked elements list 1`] = `
<div
  data-testid="list-container"
>
  <ol
    data-testid="list"
  >
    <li
      data-testid="li-0"
    >
      <span>
        {"id":0,"name":"item1"}
      </span>
    </li>
    <li
      data-testid="li-1"
    >
      <span>
        {"id":1,"name":"item2"}
      </span>
    </li>
  </ol>
</div>
`;
